h-enrichment is not working

In particular, the `burgers_ee.py` demo is broken because of this and only works if p-enrichment mode is used.

@acse-sl420 With a newly installed Firedrake, I'm now hitting errors in the adjoint tests and demos. Not sure that they are exactly the same as what you saw, but I will tackle them nonetheless.

```
FAILED test_adjoint/test_adjoint.py::test_adjoint_same_mesh[steady_flow_past_cyl-end_time] - ZeroDivisionError: float division by zero
FAILED test_adjoint/test_adjoint.py::test_adjoint_same_mesh[burgers-end_time] - AttributeError: 'FunctionSpace' object has no attribute '_ufl_is_terminal_'
FAILED test_adjoint/test_adjoint.py::test_adjoint_same_mesh[burgers-time_integrated] - AttributeError: 'FunctionSpace' object has no attribute '_ufl_is_terminal_'
FAILED test_adjoint/test_demos.py::test_demos[burgers1.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers1.py']' retur...
FAILED test_adjoint/test_demos.py::test_demos[point_discharge.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/point_discharge.py']...
FAILED test_adjoint/test_demos.py::test_demos[burgers.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers.py']' return...
FAILED test_adjoint/test_demos.py::test_demos[burgers_oo.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers_oo.py']' ret...
FAILED test_adjoint/test_demos.py::test_demos[burgers2.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers2.py']' retur...
FAILED test_adjoint/test_demos.py::test_demos[burgers_ee.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers_ee.py']' ret...
FAILED test_adjoint/test_demos.py::test_demos[burgers_time_integrated.py] - subprocess.CalledProcessError: Command '['/home/joewa/software/firedrake-parmmg/bin/python3', '/home/joewa/software/pyroteus/demos/burgers_time_integra...
```

--- a/demos/burgers1.py
+++ b/demos/burgers1.py
@@ -15,15 +15,67 @@ from pyroteus_adjoint import *
 
 # For ease, the field list and functions for obtaining the
 # function spaces, forms, solvers, and initial conditions
-# are imported from the previous demo. ::
+# are redefined as in the previous demo. ::
 
-from burgers import (
-    fields,
-    get_function_spaces,
-    get_form,
-    get_solver,
-    get_initial_condition,
-)
+fields = ["u"]
+
+
+def get_function_spaces(mesh):
+    return {"u": VectorFunctionSpace(mesh, "CG", 2)}
+
+
+def get_form(mesh_seq):
+    def form(index, solutions):
+        u, u_ = solutions["u"]
+        P = mesh_seq.time_partition
+        dt = Constant(P.timesteps[index])
+
+        # Specify viscosity coefficient
+        nu = Constant(0.0001)
+
+        # Setup variational problem
+        v = TestFunction(u.function_space())
+        F = (
+            inner((u - u_) / dt, v) * dx
+            + inner(dot(u, nabla_grad(u)), v) * dx
+            + nu * inner(grad(u), grad(v)) * dx
+        )
+        return F
+
+    return form
+
+
+def get_solver(mesh_seq):
+    def solver(index, ic):
+        function_space = mesh_seq.function_spaces["u"][index]
+        u = Function(function_space)
+
+        # Initialise 'lagged' solution
+        u_ = Function(function_space, name="u_old")
+        u_.assign(ic["u"])
+
+        # Define form
+        F = mesh_seq.form(index, {"u": (u, u_)})
+
+        # Time integrate from t_start to t_end
+        P = mesh_seq.time_partition
+        t_start, t_end = P.subintervals[index]
+        dt = P.timesteps[index]
+        t = t_start
+        while t < t_end - 1.0e-05:
+            solve(F == 0, u, ad_block_tag="u")
+            u_.assign(u)
+            t += dt
+        return {"u": u}
+
+    return solver
+
+
+def get_initial_condition(mesh_seq):
+    fs = mesh_seq.function_spaces["u"][0]
+    x, y = SpatialCoordinate(mesh_seq[0])
+    return {"u": interpolate(as_vector([sin(pi * x), 0]), fs)}
+
 
 # In line with the
 # `firedrake-adjoint demo

--- a/demos/burgers2.py
+++ b/demos/burgers2.py
@@ -12,18 +12,80 @@
 from firedrake import *
 from pyroteus_adjoint import *
 
+# Redefine the ``fields`` variable from the previous demo, as well as all the getter
+# functions. ::
+
+fields = ["u"]
+
+
+def get_function_spaces(mesh):
+    return {"u": VectorFunctionSpace(mesh, "CG", 2)}
+
+
+def get_form(mesh_seq):
+    def form(index, solutions):
+        u, u_ = solutions["u"]
+        P = mesh_seq.time_partition
+        dt = Constant(P.timesteps[index])
+
+        # Specify viscosity coefficient
+        nu = Constant(0.0001)
+
+        # Setup variational problem
+        v = TestFunction(u.function_space())
+        F = (
+            inner((u - u_) / dt, v) * dx
+            + inner(dot(u, nabla_grad(u)), v) * dx
+            + nu * inner(grad(u), grad(v)) * dx
+        )
+        return F
+
+    return form
+
+
+def get_solver(mesh_seq):
+    def solver(index, ic):
+        function_space = mesh_seq.function_spaces["u"][index]
+        u = Function(function_space)
+
+        # Initialise 'lagged' solution
+        u_ = Function(function_space, name="u_old")
+        u_.assign(ic["u"])
+
+        # Define form
+        F = mesh_seq.form(index, {"u": (u, u_)})
+
+        # Time integrate from t_start to t_end
+        P = mesh_seq.time_partition
+        t_start, t_end = P.subintervals[index]
+        dt = P.timesteps[index]
+        t = t_start
+        while t < t_end - 1.0e-05:
+            solve(F == 0, u, ad_block_tag="u")
+            u_.assign(u)
+            t += dt
+        return {"u": u}
+
+    return solver
+
+
+def get_initial_condition(mesh_seq):
+    fs = mesh_seq.function_spaces["u"][0]
+    x, y = SpatialCoordinate(mesh_seq[0])
+    return {"u": interpolate(as_vector([sin(pi * x), 0]), fs)}
+
+
+def get_qoi(mesh_seq, solutions, i):
+    def end_time_qoi():
+        u = solutions["u"]
+        return inner(u, u) * ds(2)
+
+    return end_time_qoi
+
+
 # The solver, initial condition and QoI may be imported from the
 # previous demo. The same basic setup is used. The only difference
 # is that the :class:`MeshSeq` contains two meshes. ::
-
-from burgers1 import (
-    fields,
-    get_function_spaces,
-    get_form,
-    get_solver,
-    get_initial_condition,
-    get_qoi,
-)
 
 n = 32
 meshes = [UnitSquareMesh(n, n, diagonal="left"), UnitSquareMesh(n, n, diagonal="left")]

--- a/demos/burgers_ee.py
+++ b/demos/burgers_ee.py
@@ -22,21 +22,83 @@
 # obtained by adding more degrees of freedom to the base space. This could
 # be done by solving global or local auxiliary PDEs, or by applying patch
 # recovery type methods. Currently, only global enrichment is supported in
-# Pyroteus.
-#
-# First, import the various functions and create a sequence of meshes
-# and a :class:`TimePartition`. ::
+# Pyroteus. ::
 
 from firedrake import *
 from pyroteus_adjoint import *
-from burgers1 import (
-    fields,
-    get_function_spaces,
-    get_form,
-    get_solver,
-    get_initial_condition,
-    get_qoi,
-)
+
+# Redefine the ``fields`` variable and the getter functions as in the first
+# adjoint Burgers demo. ::
+
+fields = ["u"]
+
+
+def get_function_spaces(mesh):
+    return {"u": VectorFunctionSpace(mesh, "CG", 2)}
+
+
+def get_form(mesh_seq):
+    def form(index, solutions):
+        u, u_ = solutions["u"]
+        P = mesh_seq.time_partition
+        dt = Constant(P.timesteps[index])
+
+        # Specify viscosity coefficient
+        nu = Constant(0.0001)
+
+        # Setup variational problem
+        v = TestFunction(u.function_space())
+        F = (
+            inner((u - u_) / dt, v) * dx
+            + inner(dot(u, nabla_grad(u)), v) * dx
+            + nu * inner(grad(u), grad(v)) * dx
+        )
+        return F
+
+    return form
+
+
+def get_solver(mesh_seq):
+    def solver(index, ic):
+        function_space = mesh_seq.function_spaces["u"][index]
+        u = Function(function_space)
+
+        # Initialise 'lagged' solution
+        u_ = Function(function_space, name="u_old")
+        u_.assign(ic["u"])
+
+        # Define form
+        F = mesh_seq.form(index, {"u": (u, u_)})
+
+        # Time integrate from t_start to t_end
+        P = mesh_seq.time_partition
+        t_start, t_end = P.subintervals[index]
+        dt = P.timesteps[index]
+        t = t_start
+        while t < t_end - 1.0e-05:
+            solve(F == 0, u, ad_block_tag="u")
+            u_.assign(u)
+            t += dt
+        return {"u": u}
+
+    return solver
+
+
+def get_initial_condition(mesh_seq):
+    fs = mesh_seq.function_spaces["u"][0]
+    x, y = SpatialCoordinate(mesh_seq[0])
+    return {"u": interpolate(as_vector([sin(pi * x), 0]), fs)}
+
+
+def get_qoi(mesh_seq, solutions, i):
+    def end_time_qoi():
+        u = solutions["u"]
+        return inner(u, u) * ds(2)
+
+    return end_time_qoi
+
+
+# Next, create a sequence of meshes and a :class:`TimePartition`. ::
 
 n = 32
 meshes = [UnitSquareMesh(n, n, diagonal="left"), UnitSquareMesh(n, n, diagonal="left")]

--- a/demos/burgers_time_integrated.py
+++ b/demos/burgers_time_integrated.py
@@ -11,7 +11,41 @@
 
 from firedrake import *
 from pyroteus_adjoint import *
-from burgers import get_initial_condition, get_function_spaces, get_form
+
+# Redefine the ``get_initial_condition``, ``get_function_spaces``,
+# and ``get_form`` functions as in the first Burgers demo. ::
+
+
+def get_function_spaces(mesh):
+    return {"u": VectorFunctionSpace(mesh, "CG", 2)}
+
+
+def get_form(mesh_seq):
+    def form(index, solutions):
+        u, u_ = solutions["u"]
+        P = mesh_seq.time_partition
+        dt = Constant(P.timesteps[index])
+
+        # Specify viscosity coefficient
+        nu = Constant(0.0001)
+
+        # Setup variational problem
+        v = TestFunction(u.function_space())
+        F = (
+            inner((u - u_) / dt, v) * dx
+            + inner(dot(u, nabla_grad(u)), v) * dx
+            + nu * inner(grad(u), grad(v)) * dx
+        )
+        return F
+
+    return form
+
+
+def get_initial_condition(mesh_seq):
+    fs = mesh_seq.function_spaces["u"][0]
+    x, y = SpatialCoordinate(mesh_seq[0])
+    return {"u": interpolate(as_vector([sin(pi * x), 0]), fs)}
+
 
 # The solver needs to be modified slightly in order to take
 # account of time dependent QoIs. The Burgers solver

--- a/pyroteus/quality.py
+++ b/pyroteus/quality.py
@@ -137,9 +137,9 @@ class QualityMeasure:
             c = ufl.sqrt(ufl.dot(edge3, edge3))
             detJ = ufl.JacobianDeterminant(self.mesh)
             jacobian_sign = ufl.sign(detJ)
-            max_product = ufl.Max(
-                ufl.Max(ufl.Max(a * b, a * c), ufl.Max(b * c, b * a)),
-                ufl.Max(c * a, c * b),
+            max_product = ufl.max_value(
+                ufl.max_value(ufl.max_value(a * b, a * c), ufl.max_value(b * c, b * a)),
+                ufl.max_value(c * a, c * b),
             )
             return interpolate(detJ / max_product * jacobian_sign, self.P0)
         else:

